internal/worker/server.go

```diff
@@ -267,47 +252,34 @@ func (s *Server) RequestJob(ctx context.Context, arch string, jobTypes []string)
 		}
 	}
 
-	s.runningMutex.Lock()
-	defer s.runningMutex.Unlock()
-	s.running[token] = jobId
-
 	if jobType == "osbuild:"+arch {
 		jobType = "osbuild"
 	} else if jobType == "osbuild-koji:"+arch {
 		jobType = "osbuild-koji"
 	}
 
-	return token, jobId, jobType, args, dynamicArgs, nil
-}
-
-func (s *Server) RunningJob(token uuid.UUID) (uuid.UUID, error) {
-	s.runningMutex.Lock()
-	defer s.runningMutex.Unlock()
-
-	jobId, ok := s.running[token]
-	if !ok {
-		return uuid.Nil, ErrTokenNotExist
-	}
-
-	return jobId, nil
+	return jobId, token, jobType, args, dynamicArgs, nil
 }
 
 func (s *Server) FinishJob(token uuid.UUID, result json.RawMessage) error {
-	s.runningMutex.Lock()
-	defer s.runningMutex.Unlock()
-
-	jobId, ok := s.running[token]
-	if !ok {
-		return ErrTokenNotExist
+	jobId, err := s.jobs.IdFromToken(token)
+	if err != nil {
+		switch err {
+		case jobqueue.ErrNotExist:
+			return ErrInvalidToken
+		default:
+			return err
+		}
 	}
 
-	// Always delete the running job, even if there are errors finishing
-	// the job, because callers won't call this a second time on error.
-	delete(s.running, token)
```

Comment:
Is the token correctly invalidated in the new code? Maybe make a new version of `IdFromToken()` that pops the token so that `FinishJob()` is guaranteed to return `ErrInvalidToken` if it is called more than once?

Reply:
Ah yea, I just build the map regardless if it's running or not. I'll make sure to change this, and add a test for it to make sure :)

Reply:
Hm, `ErrInvalidToken` results in a 404 `not found`. Now the error will be `ErrJobNotRunning`, which also results in a 404 `not found` :D So it'll be hard to test, but i'll make sure to remove tokens once the job has finished or has been cancelled.

Reply:
I'll not be removing the token when it has been cancelled to preserve current behavior: If a job is cancelled you can ask if it's cancelled via GetJob as long as composer isn't restarted. As soon as it is restarted, the token will not be found as the token is gone.

Reply:
Maybe add a todo about cleaning this up? Sounds like the current semantics is weird

Reply:
I'll make an issue. Tbh since it's 'internal' api between the workers and composer it's not hugely important. But I admit the behavior is a little weird.